flow/hooks/git_workspace_tracking.py

```diff
@@ -0,0 +1,115 @@
+# Copyright (c) 2018 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from __future__ import print_function
+import os
+import json
+import logging
+from collections import defaultdict
+
+import git
+
+from .git_util import collect_metadata_with_git as collect_metadata
+
+
+logger = logging.getLogger('git_tracking')
+
+GIT_COMMIT_MSG = """{title}
+
+*This commit was auto-generated.*
+"""
+
+
+_WARNINGS = defaultdict(set)
+
+
+def _ignore_hidden_files(root):
+    "Do not commit hidden files to the git repository."
+    with open(os.path.join(root, '.gitignore'), mode='a+') as file:
+        file.seek(0)
+        if '.*\n' not in file.readlines():
+            file.write('.*\n')
+
+
+def _get_or_init_git_repo(root):
+    "Return the git repository for root and initialize it if necessary."
+    try:
+        return git.Repo(root)
+    except git.exc.InvalidGitRepositoryError:
+        logger.info("Initializing git repo for '{}'".format(root))
+        _ignore_hidden_files(root)
+        return git.Repo.init(root)
+
+
+def _commit(repo, title):
+    try:
+        repo.git.commit('-m', GIT_COMMIT_MSG.format(title=title))
+    except git.exc.GitCommandError as error:
+        if "nothing to commit, working tree clean" in str(error):
+            pass
+        else:
+            raise
+
+
+def get_project_source_git_metadata(project, strict=False):
+    repo = git.Repo(project.root_directory())
+    wd_rel = os.path.relpath(project.workspace(), project.root_directory())
+    for path in repo.untracked_files:
+        if os.path.commonprefix((wd_rel, path)):
+            if project.workspace() not in _WARNINGS['workspace_not_ignored']:
+                logger.warn("Skipping workspace directory for project '{}'.".format(project))
+                _WARNINGS['workspace_not_ignored'].add(project.workspace())
+            break
+    if repo.is_dirty():
+        if strict:
+            raise RuntimeError(
+                "Unable to get project source git commit id, repository is dirty.")
+        else:
+            if project.root_directory() not in _WARNINGS['git_dirty']:
+                logger.warn("Unable to get project source git commit id, repository is dirty.")
+                _WARNINGS['git_dirty'].add(project.root_directory())
+    return {
+        'commit_id': str(repo.commit()),
+        'dirty': repo.is_dirty(),
+    }
```

Comment:
Is this function meant to be a distinct hook for users? It's not used by the `TrackWorkspaceWithGit` class so I'm not quite sure what its purpose is.